File: 61-rotate-list/61-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head==null || head.next == null) return head;
        
        ListNode dummy = new ListNode(0, head);
        ListNode fast = dummy;
        ListNode slow = dummy;
        int length=0;
        while (fast.next != null) {
            fast = fast.next;
            length += 1;
        };
        
        for(int j=length-k%length; j > 0; j--) {
            slow = slow.next;
        };
        
    fast.next=dummy.next; //Do the rotation
    dummy.next=slow.next;
    slow.next=null;
    
    return dummy.next;
        
        
    };
}

File: 61-rotate-list/61-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head: return None
        if k == 0: return head
        ans = head2 = head3 = head
        size = 1
        while head.next:
            size += 1
            head = head.next
        
        move_to_the_right_size =  k % size
        if move_to_the_right_size == 0: return ans
        for i in range(size - move_to_the_right_size - 1):
            ans = ans.next
        #print(ans)
        tmp = ans.next
        ans.next = None
        head.next = head2
        return tmp
        